tests/test_models_gpt2.py

```diff
@@ -0,0 +1,159 @@
+    cfg.freeze()
+
+    # Generate TN layout
+    cfg_tn = cfg.clone()
+    cfg_tn.defrost()
+    cfg_tn.MODEL.layout = 'TN'
+    cfg_tn.freeze()
+    
+    batch_size = 4
+    sequence_length = 16
+    ctx = mx.cpu()
+    inputs = mx.np.random.randint(0, 1000, (batch_size, sequence_length), ctx=ctx)
+
+    gpt2_model = GPT2Model.from_cfg(cfg)
+    gpt2_model.initialize(ctx=ctx)
+    gpt2_model.hybridize()
+    hiddens, _ = gpt2_model(
+        inputs,
+        gpt2_model.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
+    gpt2_model_tn = GPT2Model.from_cfg(cfg_tn)
+    gpt2_model_tn.share_parameters(gpt2_model.collect_params())
+    gpt2_model_tn.hybridize()
+    hiddens_tn, _ = gpt2_model_tn(
+        inputs.T,
+        gpt2_model_tn.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
+    assert_allclose(np.swapaxes(hiddens_tn.asnumpy(), 0, 1),
+                    hiddens.asnumpy(), 1E-4, 1E-4)
+
+    # Test for GPT2ForLM
+    gpt2_lm_model = GPT2ForLM(cfg)
+    gpt2_lm_model.initialize(ctx=ctx)
+    gpt2_lm_model.hybridize()
+    logits, states = gpt2_lm_model(
+        inputs,
+        gpt2_lm_model.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
+    gpt2_lm_model_tn = GPT2ForLM(cfg_tn)
+    gpt2_lm_model_tn.share_parameters(gpt2_lm_model.collect_params())
+    gpt2_lm_model_tn.hybridize()
+    logits_tn, states_tn = gpt2_lm_model_tn(
+        inputs.T,
+        gpt2_lm_model_tn.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
+    assert_allclose(np.swapaxes(logits_tn.asnumpy(), 0, 1),
+                    logits.asnumpy(), 1E-4, 1E-4)
+    assert_allclose(np.swapaxes(states_tn.asnumpy(), 2, 3),
+                    states.asnumpy(), 1E-4, 1E-4)
+
+
+def test_gpt2_incremental_states():
+    ctx = mx.cpu()
+    batch_size = 4
+    sequence_length = 5
+    inputs = mx.np.random.randint(0, 1000, (batch_size, sequence_length), ctx=ctx)
+
+    cfg = GPT2Model.get_cfg()
+    gpt2_model = GPT2Model.from_cfg(cfg)
+    gpt2_model.initialize(ctx=ctx)
+    gpt2_model.hybridize()
+
+    one_time_hiddens, one_time_states = gpt2_model(
+        inputs,
+        gpt2_model.init_states(batch_size, ctx),
+        mx.np.array(0, dtype=np.int32, ctx=ctx)
+    )
+    
+    states = gpt2_model.init_states(batch_size, ctx)
+    for i in range(sequence_length):
+        hiddens, states = gpt2_model(
+            inputs[:, i:i+1],
+            states,
+            mx.np.array(i, dtype=np.int32, ctx=ctx)
+        )
+    incremental_states = states
+    incremental_hiddens = hiddens
+    assert_allclose(incremental_states.asnumpy(),
+                    states.asnumpy(), 1E-4, 1E-4)
+    assert_allclose(incremental_hiddens.asnumpy(),
+                    hiddens.asnumpy(), 1E-4, 1E-4)
+
+
+#@pytest.mark.remote_required
+@pytest.mark.parametrize('model_name', list_pretrained_gpt2())
+def test_gpt2(model_name):
```

Comment:
You may refer to test_models_bert.py to add the ctx support. You can add a ctx flag to in the arguments like https://github.com/dmlc/gluon-nlp/blob/66e5e057347c3710eb8fa27c134a9303309e94c0/tests/test_models_bert.py#L94.

So you can test the model on both cpu and GPU.